### PR TITLE
Hide the current user's profile checkbox for bulk actions

### LINKED ISSUE #29838: BO > Profiles - If I bulk delete Profiles including mine, only the ones above mine will be deleted
### Prerequisites

- [x] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [x] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

https://user-images.githubusercontent.com/16019289/193547794-ad3f63fa-349b-46f5-bc1a-6f661fcb3ae3.mov

In my case, only "manager" profile is deleted, when 3 are selected.

It is not a regression.

### Expected behavior

I want to be able to delete everything except my own profile (or delete nothing, depending on PM decision)

### Steps to reproduce

As a employee with SuperAdmin Profile:
1. Go to BO > Team > Employee
2. Check All > Bulk Action > Delete selected
3. See message : "You cannot delete your own profile"

Result:
Every Profile _above_ SuperAdmin Profile are deleted
Every Profile _below_ SuperAdmin Profile are not deleted

### PrestaShop version(s) where the bug happened

1.7.7.8, 1.7.8.6, 8.0.0

### PHP version(s) where the bug happened

7.4, 8.0

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by 0x346e3730**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Hide the checkbox for the current user's profile when trying to bulk delete profiles
| Type?             | bug fix
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes https://github.com/PrestaShop/PrestaShop/issues/29838
| Related PRs       | N/A
| How to test?      | In the BackOffice, go to Advanced Parameters > Team, click the Profiles tab, check all the profiles boxes and then bulk delete them.
| Possible impacts? | Profiles

### CODE REVIEW

**Review by matks — COMMENTED** [review]
@0x346e3730 at the bottom of https://github.com/PrestaShop/PrestaShop/issues/29838 @eshraw mentioned:

"Best case scenario should be "I cannot select my own profile for a bulk action" as bulk actions (enable/disable/delete) are affecting profile integrity."

You have protected backend to make sure bulk delete does not halt if 'own profile' is deleted, can you check if you can add to this PR "I cannot select my own profile for a bulk action" ?

**Comment by kpodemski** [comment]
I like the idea of having the possibility to hide the checkbox for certain records. One idea I have, let me know what you think:
a way to provide a Grid with a list of record IDs where you want to disable bulk actions

```
 ->add((new BulkActionColumn('bulk'))
            ->setOptions([
                'bulk_field' => 'id_profile',
                'skipped_entries' => [1, 2, 3]
            ])
```

🤔

**Comment by 0x346e3730** [comment]
I was thinking about a "more generic" way of doing so. I don't know if the GridDefinition (this is where the snippet you sent would go) is the best place to do that as right now the only use case for this is dynamic based on user context regarding the data and it seems to me it's better if it's in the DataFactory side rather than in GridDefinition (especially if the logic depends on the data, it's way harder to access the data to not say impossible), but nothing is set in stone

File: src/Core/Grid/Data/Factory/ProfileGridDataFactoryDecorator.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Grid\Data\Factory;
```
Comment (matks) [inline src/Core/Grid/Data/Factory/ProfileGridDataFactoryDecorator.php:29]: For new classes you can add declare strict types

https://devdocs.prestashop-project.org/1.7/development/coding-standards/#strict-typing

**Comment by Progi1984** [comment]
Hi @0x346e3730, I have implemeneted the option`disabled_field` in Grids and specifically inOrderReturnStates Grid. My idea is that you can use it : implemened it in the factory : https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Definition/Factory/OrderReturnStatesGridDefinitionFactory.php#L100 and base it on a virtual field `isCurrentUser` (for example) : https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Grid/Query/OrderQueryBuilder.php#L95 (example of a non existing field in database).

**Comment by 0x346e3730** [comment]
Hey @Progi1984, the answer is kind of the same as the one I did to Krystian : because the rendering is dynamic based on the user context it cannot be done in the GridDefinition part (or I misunderstood the code you linked me). Bump @PrestaShop/prestashop-maintainers to get another review round as this PR was stalled and forgotten

File: src/Core/Grid/Data/Factory/ProfileGridDataFactoryDecorator.php (hunk 1 above)
Comment (0x346e3730) [reply src/Core/Grid/Data/Factory/ProfileGridDataFactoryDecorator.php:29]: Done

**Review by kpodemski — CHANGES_REQUESTED** [review]
Hi @0x346e3730 

Sorry for the late review :/ Tests are failing, so I think it would be better for you to re-check PR and I'll get back to it as soon as you ping me!

**Comment by 0x346e3730** [comment]
Updated with develop to re-trigger the CI because the logs were too old.

**Comment by 0x346e3730** [comment]
CI is ✅ cc @kpodemski

File: src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml [hunk 2]
```diff
@@ -186,6 +186,12 @@ services:
       - '@prestashop.core.grid.query.doctrine_query_parser'
       - 'profile'
 
+  prestashop.core.grid.data_factory.profiles_decorator:
+    class: 'PrestaShop\PrestaShop\Core\Grid\Data\Factory\ProfileGridDataFactoryDecorator'
+    arguments:
+      - '@prestashop.core.grid.data_factory.profiles'
+      - '@security.helper'
+
```
Comment (lartist) [inline src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml]: Hi @0x346e3730, can you plz feat with https://github.com/PrestaShop/ADR/blob/master/0019-FQCN-Autowiring.md

1. create the FQCN version of '@prestashop.core.grid.data_factory.profiles' and '@security.helper' and also aliases version (for BC compatibility)
2. Declare the new service like this:
```php
PrestaShop\PrestaShop\Core\Grid\Data\Factory\ProfileGridDataFactoryDecorato:
    autowire: true
    public: false
```

**Review by matthieu-rolland — COMMENTED** [review]
Nice PR :+1:  

Edit: my mistake, I was checking in "employee" and not "profile".

**Review by matthieu-rolland — DISMISSED** [review]
LGTM :+1:  (minus @lartist FQCN feedback)

File: src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml (hunk 2 above)
Comment (0x346e3730) [reply src/PrestaShopBundle/Resources/config/services/core/grid/grid_data_factory.yml]: Done

**Review by matks — APPROVED** [review]

**Review by kpodemski — APPROVED** [review]

**Comment by kpodemski** [comment]
QA: as the scenario is simple, you can test it even when tests are still running, I had to restart them

**Review by djoelleuch — APPROVED** [review]
Hello @0x346e3730 ,

I have tested your PR and it's well fixed !

[Untitled_ Jul 28, 2023 12_06 PM.webm](https://github.com/PrestaShop/PrestaShop/assets/115152460/488547fc-7f6b-4168-acfb-ec1f9e8ee581)

automated tests : OK (https://github.com/aniszr/ga.tests.ui.pr/actions/runs/5689496314)
It's QA approved ✔️
Thank you :rocket:

**Comment by kpodemski** [comment]
thanks @0x346e3730

**Pull request merged**